**Importing the Required Libraries**

In [ ]:
import pandas as pd
import numpy as np

bdata = pd.read_csv("Building_Permits.csv")

[link text](https://)**Printing the Head of the Dataset**

In [ ]:
bdata.head()

Permit Number  Permit Type            Permit Type Definition  \
0   2.01604E+11            4                      sign - erect   
1   2.01605E+11            3  additions alterations or repairs   
2   2.01611E+11            8            otc alterations permit   
3   2.01611E+11            6                       demolitions   
4   2.01706E+11            8            otc alterations permit   

  Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0            4/19/2016   306    7            440                  NaN   
1            5/27/2016   595  203           1647                  NaN   
2            11/7/2016   156   11           1230                  NaN   
3           11/28/2016   342    1            950                  NaN   
4            6/14/2017  4105    9            800                  NaN   

  Street Name Street Suffix  ...  Existing Construction Type  \
0       Geary            St  ...                         3.0   
1     Pacific            Av  ...                         1.0   
2     Pacific            Av  ...                         5.0   
3      Market            St  ...                         3.0   
4     Indiana            St  ...                         1.0   

  Existing Construction Type Description Proposed Construction Type  \
0                          constr type 3                        NaN   
1                          constr type 1                        1.0   
2                         wood frame (5)                        5.0   
3                          constr type 3                        NaN   
4                          constr type 1                        1.0   

  Proposed Construction Type Description Site Permit Supervisor District  \
0                                    NaN         NaN                   3   
1                          constr type 1         NaN                   3   
2                         wood frame (5)         NaN                   3   
3                                    NaN         NaN                   6   
4                          constr type 1         NaN                  10   

  Neighborhoods - Analysis Boundaries Zipcode  \
0                          Tenderloin   94102   
1                        Russian Hill   94109   
2                            Nob Hill   94109   
3                          Tenderloin   94102   
4                        Potrero Hill   94107   

                                   Location     Record ID  
0  (37.78733980600732, -122.41063199757738)  1.420160e+12  
1   (37.7946573324287, -122.42232562979227)  1.424860e+12  
2  (37.79595867909168, -122.41557405519474)  1.443570e+12  
3  (37.78315261897309, -122.40950883997789)  1.445480e+11  
4  (37.75922331346539, -122.39170402628598)  1.466910e+12  

[5 rows x 43 columns]

**Calculating the missing points that the dataset have**

In [ ]:
miss=bdata.isnull().sum()
miss[0:]

Permit Number                               0
Permit Type                                 0
Permit Type Definition                      0
Permit Creation Date                        0
Block                                       0
Lot                                         0
Street Number                               0
Street Number Suffix                      144
Street Name                                 0
Street Suffix                               1
Unit                                      119
Unit Suffix                               144
Description                                 0
Current Status                              0
Current Status Date                         0
Filed Date                                  0
Issued Date                                24
Completed Date                            111
First Construction Document Date           24
Structural Notification                   143
Number of Existing Stories                 75
Number of Proposed Stories        

**Calculating missing percentage**

In [ ]:
total = np.product(bdata.shape)
totals = miss.sum()


percent_missing = (totals/total) * 100
print(percent_missing)

37.337610264635124


**Here the columns that containing above 143 missing values were dropped**

In [ ]:
columns_with_na_dropped = bdata.dropna(axis=1, thresh=bdata.shape[0] - 141)
columns_with_na_dropped.head()

Permit Number  Permit Type            Permit Type Definition  \
0   2.01604E+11            4                      sign - erect   
1   2.01605E+11            3  additions alterations or repairs   
2   2.01611E+11            8            otc alterations permit   
3   2.01611E+11            6                       demolitions   
4   2.01706E+11            8            otc alterations permit   

  Permit Creation Date Block  Lot  Street Number Street Name Street Suffix  \
0            4/19/2016   306    7            440       Geary            St   
1            5/27/2016   595  203           1647     Pacific            Av   
2            11/7/2016   156   11           1230     Pacific            Av   
3           11/28/2016   342    1            950      Market            St   
4            6/14/2017  4105    9            800     Indiana            St   

   Unit  ... Plansets Existing Construction Type  \
0   0.0  ...      2.0                        3.0   
1   NaN  ...      2.0                        1.0   
2   0.0  ...      2.0                        5.0   
3   NaN  ...      2.0                        3.0   
4   NaN  ...      2.0                        1.0   

  Existing Construction Type Description Proposed Construction Type  \
0                          constr type 3                        NaN   
1                          constr type 1                        1.0   
2                         wood frame (5)                        5.0   
3                          constr type 3                        NaN   
4                          constr type 1                        1.0   

  Proposed Construction Type Description Supervisor District  \
0                                    NaN                   3   
1                          constr type 1                   3   
2                         wood frame (5)                   3   
3                                    NaN                   6   
4                          constr type 1                  10   

  Neighborhoods - Analysis Boundaries  Zipcode  \
0                          Tenderloin    94102   
1                        Russian Hill    94109   
2                            Nob Hill    94109   
3                          Tenderloin    94102   
4                        Potrero Hill    94107   

                                   Location     Record ID  
0  (37.78733980600732, -122.41063199757738)  1.420160e+12  
1   (37.7946573324287, -122.42232562979227)  1.424860e+12  
2  (37.79595867909168, -122.41557405519474)  1.443570e+12  
3  (37.78315261897309, -122.40950883997789)  1.445480e+11  
4  (37.75922331346539, -122.39170402628598)  1.466910e+12  

[5 rows x 37 columns]

**Here all the NAN values are generalized to 0**

In [ ]:
columns_with_na_dropped.fillna(0)

Permit Number  Permit Type            Permit Type Definition  \
0     2.01604E+11            4                      sign - erect   
1     2.01605E+11            3  additions alterations or repairs   
2     2.01611E+11            8            otc alterations permit   
3     2.01611E+11            6                       demolitions   
4     2.01706E+11            8            otc alterations permit   
..            ...          ...                               ...   
140   2.01706E+11            3  additions alterations or repairs   
141   2.01706E+11            8            otc alterations permit   
142   2.01706E+11            8            otc alterations permit   
143       M803330            8            otc alterations permit   
144   2.01706E+11            8            otc alterations permit   

    Permit Creation Date Block  Lot  Street Number             Street Name  \
0              4/19/2016   306    7            440                   Geary   
1              5/27/2016   595  203           1647                 Pacific   
2              11/7/2016   156   11           1230                 Pacific   
3             11/28/2016   342    1            950                  Market   
4              6/14/2017  4105    9            800                 Indiana   
..                   ...   ...  ...            ...                     ...   
140            6/27/2017    97   18           1178                   Union   
141            6/27/2017   297   28            515                   Mason   
142            6/29/2017  8721   32            455  Mission Bay Blvd South   
143            6/29/2017  3733  161            320              Clementina   
144            6/30/2017  2829   23            859               Elizabeth   

    Street Suffix  Unit  ... Plansets Existing Construction Type  \
0              St   0.0  ...      2.0                        3.0   
1              Av   0.0  ...      2.0                        1.0   
2              Av   0.0  ...      2.0                        5.0   
3              St   0.0  ...      2.0                        3.0   
4              St   0.0  ...      2.0                        1.0   
..            ...   ...  ...      ...                        ...   
140            St   0.0  ...      2.0                        5.0   
141            St   0.0  ...      2.0                        1.0   
142             0   0.0  ...      2.0                        1.0   
143            St   0.0  ...      0.0                        0.0   
144            St   0.0  ...      0.0                        5.0   

    Existing Construction Type Description Proposed Construction Type  \
0                            constr type 3                        0.0   
1                            constr type 1                        1.0   
2                           wood frame (5)                        5.0   
3                            constr type 3                        0.0   
4                            constr type 1                        1.0   
..                                     ...                        ...   
140                         wood frame (5)                        5.0   
141                          constr type 1                        1.0   
142                          constr type 1                        1.0   
143                                      0                        0.0   
144                         wood frame (5)                        5.0   

    Proposed Construction Type Description Supervisor District  \
0                                        0                   3   
1                            constr type 1                   3   
2                           wood frame (5)                   3   
3                                        0                   6   
4                            constr type 1                  10   
..                                     ...                 ...   
140                         wood frame (5)                   2   
141                         

**Updated missing values**
Here the missing values are removed based on the dependency and usability

In [ ]:
miss=columns_with_na_dropped.isnull().sum()
miss[0:]

Permit Number                               0
Permit Type                                 0
Permit Type Definition                      0
Permit Creation Date                        0
Block                                       0
Lot                                         0
Street Number                               0
Street Name                                 0
Street Suffix                               1
Unit                                      119
Description                                 0
Current Status                              0
Current Status Date                         0
Filed Date                                  0
Issued Date                                24
Completed Date                            111
First Construction Document Date           24
Number of Existing Stories                 75
Number of Proposed Stories                 83
Fire Only Permit                          138
Permit Expiration Date                     93
Estimated Cost                    

**Missing percentage reduced**

In [ ]:
total = np.product(bdata.shape)
totals = miss.sum()


percent_missing = (totals/total) * 100
print(percent_missing)

23.496391339214114


In [ ]:
df = pd.DataFrame(bdata)

In [ ]:
print(df)

    Permit Number  Permit Type            Permit Type Definition  \
0     2.01604E+11            4                      sign - erect   
1     2.01605E+11            3  additions alterations or repairs   
2     2.01611E+11            8            otc alterations permit   
3     2.01611E+11            6                       demolitions   
4     2.01706E+11            8            otc alterations permit   
..            ...          ...                               ...   
140   2.01706E+11            3  additions alterations or repairs   
141   2.01706E+11            8            otc alterations permit   
142   2.01706E+11            8            otc alterations permit   
143       M803330            8            otc alterations permit   
144   2.01706E+11            8            otc alterations permit   

    Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0              4/19/2016   306    7            440                  NaN   
1              5/27/2016   595  2

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [ ]:
X = df[['Permit Type']]  # Features
y = df['Street Number']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [ ]:
model = LinearRegression()


In [ ]:
model.fit(X_train, y_train)


LinearRegression()

In [ ]:
y_pred = model.predict(X_test)


In [ ]:
r_squared = model.score(X_test, y_test)


In [ ]:
r_squared = model.score(X_test, y_test)


In [ ]:
print("R-squared:", r_squared)


R-squared: -0.020267976377800867


So the regression performed poorly
